include/phasar/DataFlow/PathSensitivity/PathSensitivityManagerBase.h

```diff
@@ -0,0 +1,190 @@
+class Instruction;
+} // namespace llvm
+
+namespace psr {
+
+template <typename Derived, typename AnalysisDomainTy, typename GraphType>
+class PathSensitivityManagerMixin;
+
+template <typename N> class PathSensitivityManagerBase {
+public:
+  using n_t = N;
+  using graph_type = AdjacencyList<llvm::SmallVector<n_t, 0>>;
+
+  static_assert(std::is_integral_v<typename GraphTraits<graph_type>::vertex_t>);
+
+  template <typename Derived, typename AnalysisDomainTy, typename GraphType>
+  friend class PathSensitivityManagerMixin;
+
+protected:
+  using graph_traits_t = GraphTraits<graph_type>;
+  using vertex_t = typename graph_traits_t::vertex_t;
+
+private:
+  [[nodiscard]] bool assertIsDAG(const graph_type &Dag) const {
+    llvm::BitVector Visited(graph_traits_t::size(Dag));
+    llvm::DenseSet<vertex_t> CurrPath;
+    CurrPath.reserve(graph_traits_t::size(Dag));
+
+    // NOLINTNEXTLINE(readability-identifier-naming)
+    auto doAssertIsDAG = [&CurrPath, &Visited, &Dag](auto &doAssertIsDAG,
+                                                     vertex_t Vtx) {
+      if (!CurrPath.insert(Vtx).second) {
+        PHASAR_LOG_LEVEL(ERROR, "DAG has circle: Vtx: " << uintptr_t(Vtx));
+        return false;
+      }
+
+      scope_exit CurrPathPop = [&CurrPath, Vtx] { CurrPath.erase(Vtx); };
+      if (Visited.test(Vtx)) {
+        /// We have already analyzed this node
+        /// NOTE: We must check this _after_ doing the circle check. Otherwise,
+        /// that can never be true
+        return true;
+      }
+
+      Visited.set(Vtx);
+
+      for (auto Succ : graph_traits_t::outEdges(Dag, Vtx)) {
+        if (!doAssertIsDAG(doAssertIsDAG, Succ)) {
+          return false;
+        }
+      }
+      return true;
+    };
+
+    for (auto Rt : graph_traits_t::roots(Dag)) {
+      if (!doAssertIsDAG(doAssertIsDAG, Rt)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  template <typename VertexTransform>
+  [[nodiscard]] static graph_type
+  reverseDAG(graph_type &&Dag, const VertexTransform &Equiv, size_t EquivSize,
+             size_t MaxDepth) {
+    using graph_traits_t = psr::GraphTraits<graph_type>;
+    using vertex_t = typename graph_traits_t::vertex_t;
+    graph_type Ret{};
+    if constexpr (psr::is_reservable_graph_trait_v<graph_traits_t>) {
+      graph_traits_t::reserve(Ret, EquivSize);
+    }
+
+    // Allocate a buffer for the temporary data needed.
+    // We need:
+    // - A cache of vertex_t
+    // - One worklist WLConsume of pair<vertex_t, vertex_t> where we read from
+    // - One worklist WLInsert of same type where we insert to
+
+    // We iterate over the graph in levels. Each level corresponds to the
+    // distance from a root node. We always have all nodes from a level inside a
+    // worklist
+    // -- The level we are processing is in WLConsume, the next level in
+    // WLInsert. This way, we can stop the process, when we have reached the
+    // MaxDepth
+
+    auto NumBytes =
```

Comment:
in the discussion we believed that this never overflows since EquivSize is limited to the number of DAG nodes. Let's make sure this is correct. Maybe consider a check to be 100% sure.

Reply:
Added an assert